docs/dep/dep-004.md

```diff
@@ -0,0 +1,146 @@
+
+The first stage of the lifetime of the system consists of the deployment of
+Tasks on the clusterand that should be achieved through Helm packages, allowing
+the system operator to deploy and modify the configuration in whatever way they
+deem necessary.
+
+An example of why we need to do this is a security teams who wishes to have a
+custom base image forall their Tasks, which requires them to build all the
+images from scratch and push them into aprivate container registry. The host of
+the registry is going to be supplied as a parameter to Helm during deployment,
+but it should not be required when deploying a Pipeline. A Pipeline that
+discovers with a high probability vulnerabilities in a Javascript codebase
+should be able to run inany cluster as long as the Tasks are available.
+Packaging all the Tasks into one Helm packagemakes this trivial to achieve,
+since it can be easily uploaded and distributed to an OCI registry.
+
+The second stage begins as soon as the `draconctl` is invoked to deploy a
+Pipeline. The Pipeline lists a set of components. The user has 3 different
+formats to describe a component:
+
+1. a local filesystem path
+
+This is used by the system to, on the fly read a Task and deploy it on the
+cluster. There are not going to be any checks if the Task already exists, or if
+the version of the Task is newer than the Task being fetched. It's going to be
+`applied` on the cluster, as if the user is invoking `kubectl apply ...`.
+
+2. a URL
+
+The URL is going to be resolved and the Task is going to be fetched and
+deployed on the cluster similar to the case where we have a file in the local
+filesystem.
+
+3. a pURL-inspired id of a Task deployed using Helm
+
+As mentioned previously, we wish to package all our Tasks in one Helm package
+that can be easily deployed on a cluster and provides access to all our
+components. When a user wants to reference a Task deployed via Helm on the
+cluster in a Pipeline description, they will have to use a new URI type that is
+inspired by `pURL` and looks like this: `helm://{{ helm_release_name }}/name`.
+
+The `helm_release_name` is the name of the release chosen by the operator when
+deploying the package, so it's specific to the deployment context. The name of
+the Task will be the actual name of the object. The system will discover all
+the Tasks belonging to the Helm package by filtering with the label
+`app.kubernetes.io/managed-by: Helm` and then the label
+`app.kubernetes.io/instance: { helm_release_name }` and will expect to find the
+Task. If the Task is not found, an error will be returned.
+
+## Implementation
+
+A new interface will be added to the system that can be used for an actual
+implementation but also for mocks for testing purposes:
+
+```golang
+type Applier struct {
+    Apply()
+}
+
+type Component struct {
+    ID() string
+
+    TektonV1Beta1Task() (*tektonv1beta1api.Task, error)
+
+    ManagedByHelm() bool
```

Comment:
i think ALL components should be managed by helm, this simplifies things and allows us to have all components as helm packages

Reply:
I added this so that a user can have an easy way to deploy a Task from the local repo. suppose that you are hacking away trying to get something to work. if you have to package it in Helm and deploy it that adds a couple of steps. allowing ppl to just fetch files from the filesystem or from a URL like we do now should make development a smoother process

Reply:
sure, then makes sense, trying to keep it simple but since helm allows for the installation of local packages why don't we put both components through a `helm upgrade` but in the case of a purl we resolve the purl first?

Reply:
we would need to also build the local packages and make sure that the version of the local chart is greater than the one deployed so that the `upgrade` command can behave as we expect. that's why I think it's a bit fussy and better left for the future if needed